File: library/src/test/java/me/proxer/library/api/DelimitedStringSetAdapterFactoryTest.java
package me.proxer.library.api;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;
import lombok.Value;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.lang.reflect.ParameterizedType;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author Ruben Gees
 */
class DelimitedStringSetAdapterFactoryTest {

    private DelimitedStringSetAdapterFactory factory;
    private Moshi moshi;

    @BeforeEach
    void setUp() {
        factory = new DelimitedStringSetAdapterFactory();
        moshi = new Moshi.Builder()
                .add(factory)
                .build();
    }

    @Test
    void testCreate() throws NoSuchFieldException {
        final DelimitedStringSet annotation = StringSetTestClass.class.getField("value")
                .getAnnotation(DelimitedStringSet.class);

        final ParameterizedType type = Types.newParameterizedType(Set.class, String.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNotNull();
    }

    @Test
    void testCreateInvalidType() throws NoSuchFieldException {
        final DelimitedStringSet annotation = StringSetTestClass.class.getField("value")
                .getAnnotation(DelimitedStringSet.class);

        final ParameterizedType type = Types.newParameterizedType(List.class, String.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testCreateInvalidParameterType() throws NoSuchFieldException {
        final DelimitedStringSet annotation = StringSetTestClass.class.getField("value")
                .getAnnotation(DelimitedStringSet.class);

        final ParameterizedType type = Types.newParameterizedType(Set.class, Boolean.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testCreateNoParameterType() throws NoSuchFieldException {
        final DelimitedStringSet annotation = StringSetTestClass.class.getField("value")
                .getAnnotation(DelimitedStringSet.class);

        final JsonAdapter<?> adapter = factory.create(Set.class, Collections.singleton(annotation), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testCreateNoAnnotation() {
        final ParameterizedType type = Types.newParameterizedType(Set.class, String.class);
        final JsonAdapter<?> adapter = factory.create(type, Collections.emptySet(), moshi);

        assertThat(adapter).isNull();
    }

    @Test
    void testFromJsonSingle() throws IOException {
        final StringSetTestClass result = moshi.adapter(StringSetTestClass.class)
                .fromJson("{\"value\":\"a\"}");

        assertThat(result).isNotNull();
        assertThat(result.value).containsExactly("a");
    }

    @Test
    void testFromJsonMultiple() throws IOException {
        final StringSetTestClass result = moshi.adapter(StringSetTestClass.class)
                .fromJson("{\"value\":\"a b\"}");

        assertThat(result).isNotNull();
        assertThat(result.value).containsExactly("a", "b");
    }

    @Test
    void testFromJsonMultipleSame() throws IOException {
        final StringSetTestClass result = moshi.adapter(StringSetTestClass.class)
                .fromJson("{\"value\":\"a b b\"}");

        assertThat(result).isNotNull();
        assertThat(result.value).containsExactly("a", "b");
    }

    @Test
    void testFromJsonEmpty() throws IOException {
        final StringSetTestClass result = moshi.adapter(StringSetTestClass.class)
                .fromJson("{\"value\":\"\"}");

        assertThat(result).isNotNull();
        assertThat(result.value).isEmpty();
    }

    @Test
    void testFromJsonNull() throws IOException {
        final StringSetTestClass result = moshi.adapter(StringSetTestClass.class)
                .fromJson("{\"value\":null}");

        assertThat(result).isNotNull();
        assertThat(result.value).isEmpty();
    }

    @Test
    void testFromJsonDifferentDelimiter() throws IOException {
        final StringSetWithDelimiterTestClass result = moshi.adapter(StringSetWithDelimiterTestClass.class)
                .fromJson("{\"value\":\"a, b,c\"}");

        assertThat(result).isNotNull();
        assertThat(result.value).containsExactly("a", "b,c");
    }

    @Test
    void testFromJsonValuesToKeep() throws IOException {
        final StringSetWithValuesToKeepTestClass result = moshi.adapter(StringSetWithValuesToKeepTestClass.class)
                .fromJson("{\"value\":\"a b c\"}");

        assertThat(result).isNotNull();
        assertThat(result.value).containsExactly("a b", "c");
    }

    @Test
    void testToJsonSingle() {
        final StringSetTestClass subject = new StringSetTestClass(Collections.singleton("a"));
        final String result = moshi.adapter(StringSetTestClass.class).toJson(subject);

        assertThat(result).isEqualTo("{\"value\":\"a\"}");
    }

    @Test
    void testToJsonMultiple() {
        final StringSetTestClass subject = new StringSetTestClass(new HashSet<>(Arrays.asList("a", "b", "c")));
        final String result = moshi.adapter(StringSetTestClass.class).toJson(subject);

        assertThat(result).isEqualTo("{\"value\":\"a b c\"}");
    }

    @Test
    void testToJsonDifferentDelimiter() {
        final HashSet<String> values = new HashSet<>(Arrays.asList("a", "b", "c"));
        final StringSetWithDelimiterTestClass subject = new StringSetWithDelimiterTestClass(values);
        final String result = moshi.adapter(StringSetWithDelimiterTestClass.class).toJson(subject);

        assertThat(result).isEqualTo("{\"value\":\"a, b, c\"}");
    }

    @Value
    private static class StringSetTestClass {

        @DelimitedStringSet
        public Set<String> value;
    }

    @Value
    private static class StringSetWithDelimiterTestClass {

        @DelimitedStringSet(delimiter = ", ")
        public Set<String> value;
    }

    @Value
    private static class StringSetWithValuesToKeepTestClass {

        @DelimitedStringSet(valuesToKeep = "a b")
        public Set<String> value;
    }
}

File: library/src/test/kotlin/me/proxer/library/api/DelimitedStringSetAdapterFactoryTest.kt
package me.proxer.library.api

import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

/**
 * @author Ruben Gees
 */
class DelimitedStringSetAdapterFactoryTest {

    private val factory = DelimitedStringSetAdapterFactory()

    private val moshi = Moshi.Builder()
        .add(factory)
        .build()

    private val annotation = StringSetTestClass::class.java.getDeclaredField("value")
        .getAnnotation(DelimitedStringSet::class.java)

    @Test
    fun testCreate() {
        val type = Types.newParameterizedType(Set::class.java, String::class.java)
        val adapter = factory.create(type, setOf(annotation), moshi)

        assertThat(adapter).isNotNull
    }

    @Test
    fun testCreateInvalidType() {
        val type = Types.newParameterizedType(List::class.java, String::class.java)
        val adapter = factory.create(type, setOf(annotation), moshi)

        assertThat(adapter).isNull()
    }

    @Test
    fun testCreateInvalidParameterType() {
        val type = Types.newParameterizedType(Set::class.java, Boolean::class.javaObjectType)
        val adapter = factory.create(type, setOf(annotation), moshi)

        assertThat(adapter).isNull()
    }

    @Test
    fun testCreateNoParameterType() {
        val adapter = factory.create(Set::class.java, setOf(annotation), moshi)

        assertThat(adapter).isNull()
    }

    @Test
    fun testCreateNoAnnotation() {
        val type = Types.newParameterizedType(Set::class.java, String::class.java)
        val adapter = factory.create(type, emptySet(), moshi)

        assertThat(adapter).isNull()
    }

    @Test
    fun testFromJsonSingle() {
        val result = moshi.adapter(StringSetTestClass::class.java).fromJson("{\"value\":\"a\"}")

        assertThat(result).isNotNull
        assertThat(result?.value).containsExactly("a")
    }

    @Test
    fun testFromJsonMultiple() {
        val result = moshi.adapter(StringSetTestClass::class.java).fromJson("{\"value\":\"a b\"}")

        assertThat(result).isNotNull
        assertThat(result?.value).containsExactly("a", "b")
    }

    @Test
    fun testFromJsonMultipleSame() {
        val result = moshi.adapter(StringSetTestClass::class.java).fromJson("{\"value\":\"a b b\"}")

        assertThat(result).isNotNull
        assertThat(result?.value).containsExactly("a", "b")
    }

    @Test
    fun testFromJsonEmpty() {
        val result = moshi.adapter(StringSetTestClass::class.java).fromJson("{\"value\":\"\"}")

        assertThat(result).isNotNull
        assertThat(result?.value).isEmpty()
    }

    @Test
    fun testFromJsonNull() {
        val result = moshi.adapter(StringSetTestClass::class.java).fromJson("{\"value\":null}")

        assertThat(result).isNotNull
        assertThat(result?.value).isEmpty()
    }

    @Test
    fun testFromJsonDifferentDelimiter() {
        val result = moshi.adapter(StringSetWithDelimiterTestClass::class.java).fromJson("{\"value\":\"a, b,c\"}")

        assertThat(result).isNotNull
        assertThat(result?.value).containsExactly("a", "b,c")
    }

    @Test
    fun testFromJsonValuesToKeep() {
        val result = moshi.adapter(StringSetWithValuesToKeepTestClass::class.java).fromJson("{\"value\":\"a b c\"}")

        assertThat(result).isNotNull
        assertThat(result?.value).containsExactlyInAnyOrder("a b", "c")
    }

    @Test
    fun testToJsonSingle() {
        val subject = StringSetTestClass(setOf("a"))
        val result = moshi.adapter(StringSetTestClass::class.java).toJson(subject)

        assertThat(result).isEqualTo("{\"value\":\"a\"}")
    }

    @Test
    fun testToJsonMultiple() {
        val subject = StringSetTestClass(setOf("a", "b", "c"))
        val result = moshi.adapter(StringSetTestClass::class.java).toJson(subject)

        assertThat(result).isEqualTo("{\"value\":\"a b c\"}")
    }

    @Test
    fun testToJsonDifferentDelimiter() {
        val subject = StringSetWithDelimiterTestClass(setOf("a", "b", "c"))
        val result = moshi.adapter(StringSetWithDelimiterTestClass::class.java).toJson(subject)

        assertThat(result).isEqualTo("{\"value\":\"a, b, c\"}")
    }

    private data class StringSetTestClass(
        @field:DelimitedStringSet val value: Set<String>
    )

    private data class StringSetWithDelimiterTestClass(
        @field:DelimitedStringSet(delimiter = ", ") val value: Set<String>
    )

    private data class StringSetWithValuesToKeepTestClass(
        @field:DelimitedStringSet(valuesToKeep = ["a b"]) val value: Set<String>
    )
}
